instrumentation/finagle-http-23.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/finagle/v23_11/FinagleClientExtension.java

```diff
@@ -0,0 +1,132 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+public class FinagleClientExtension implements BeforeEachCallback, AfterEachCallback {
+  private ExecutorService executor;
+  private EventLoopGroup eventLoopGroup;
+  private Map<URI, Service<Request, Response>> clients;
+
+  private final Function<Http.Client, Http.Client> clientBuilder;
+
+  public FinagleClientExtension(Function<Http.Client, Http.Client> clientBuilder) {
+    this.clientBuilder = clientBuilder;
+  }
+
+  public static FinagleClientExtension http_1() {
+    return new FinagleClientExtension(Http.Client::withNoHttp2);
+  }
+
+  // todo implement http/2-specific tests
+  //    .withHttp2()
+  //    .configured(PriorKnowledge.apply(true).mk())
+  public static FinagleClientExtension http_2() {
+    return new FinagleClientExtension(Http.Client::withHttp2);
+  }
+
+  static int safePort(URI uri) {
+    int port = uri.getPort();
+    if (port == -1) {
+      port = uri.getScheme().equals("https") ? 443 : 80;
+    }
+    return port;
+  }
+
+  static Request buildRequest(String method, URI uri, Map<String, String> headers) {
+    Request request =
+        Request.apply(
+            Method.apply(method.toUpperCase(Locale.ENGLISH)),
+            uri.getPath() + (uri.getQuery() == null ? "" : "?" + uri.getRawQuery()));
+    request.host(uri.getHost() + ":" + safePort(uri));
+    headers.forEach((key, value) -> request.headerMap().put(key, value));
+    return request;
+  }
+
+  public Http.Client builderFor(URI uri) {
+    Http.Client client =
+        clientBuilder
+            .apply(Http.client())
+            .withTransport()
+            .readTimeout(Duration.fromMilliseconds(READ_TIMEOUT.toMillis()))
+            .withTransport()
+            .connectTimeout(Duration.fromMilliseconds(CONNECTION_TIMEOUT.toMillis()))
+            // the default WorkerEventLoop.Global uses cached threads which get created and
+            // therefore never release their scopes (by default)
+            .configured(new WorkerPool(eventLoopGroup).mk())
+            // disable automatic retries -- retries will result in under-counting traces in the
+            // tests
+            .withRetryBudget(RetryBudget.Empty());
+
+    if (uri.getScheme().equals("https")) {
+      client = client.withTransport().tlsWithoutValidation();
+    }
+
+    return client;
+  }
+
+  public Service<Request, Response> clientFor(URI uri) {
+    return clients.computeIfAbsent(
+        uri, ignored -> builderFor(uri).newService(uri.getHost() + ":" + safePort(uri)));
+  }
+
+  @Override
+  public void beforeEach(ExtensionContext context) {
+    // finagle uses a global event loop group by default which never gets shut down;
```

Comment:
@dmarkwat I want to look at this a bit more tomorrow. Scope leaks may be an issue in the instrumentation.

Reply:
@dmarkwat I pushed a change into this PR that hopefully resolves the context leaks. I think now you can replace `FinagleClientExtension` with a more traditional approach where you create 1 http client at the start and use it in all tests and shut down at the end. You can also remove `.configured(new WorkerPool(eventLoopGroup).mk())` from the http client configuration. You may need to create separate http clients for ssl and read timeout test (to avoid non read timeout tests getting timeouts under heavy load).

Reply:
Thanks for taking a look! I always wondered why the scopes were leaking, and your changes appear to solve the issue. I've removed those workerpool configs and now I'll start in on the test updates.

If you have a good example of a more traditional approach for handling this handy, :pray: do LMK. I modeled this after netty's own as best I could but if there's a better example I'll follow that. Otherwise, I'll report back once I've made progress.

Reply:
I think the server tests are fine as they are. The apache http client test wasn't a good sample. I only meant it to illustrate that you could have 2 http client instances that you could use depending on the request path. You don't really need any of the abstract classes or weird nested classes. I guess it should be enough to just have 1 test class extends `AbstractHttpClientTest`.

Reply:
Gotcha :ok_hand: . I thought that might have been the case as I picked apart the apache client test. It _was_ a good example for that case, though, fwiw!

All updated now :+1: